### PR TITLE
Particle BCs without relocatable device code

### PR DESCRIPTION
**Pull request by brryan**
## PR Summary

Particles currently use relocatable device code via inheritance to provide boundary condition functions on a per-particle basis. This PR moves to boundary conditions on particles via kernel launches in a manner analogous to field boundary conditions. 

Breaking: completely new particle BCs for downstream codes doing custom implementations.

## PR Checklist

- [x] Code passes cpplint
- [x] New features are documented.
- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Code is formatted
- [x] Changes are summarized in CHANGELOG.md
- [x] Change is breaking (API, behavior, ...)
  - [x] Change is *additionally* added to CHANGELOG.md in the breaking section
  - [x] PR is marked as breaking
  - [ ] Short summary API changes at the top of the PR (plus optionally with an automated update/fix script)
- [x] CI has been triggered on [Darwin](https://re-git.lanl.gov/eap-oss/parthenon/-/pipelines) for performance regression tests.
- [x] Docs build
- [x] (@lanl.gov employees) Update copyright on changed files
- [x] Merge #1009
- [x] Fix particle tracers bug (missing dependence for `Defrag` task)
- [x] Add `particle_tracers` example to CI

### CODE REVIEW

File: src/interface/swarm.cpp [hunk 1]
```diff
@@ -81,7 +81,7 @@ Swarm::Swarm(const std::string &label, const Metadata &metadata, const int nmax_
   Add("y", Metadata({Metadata::Real}));
   Add("z", Metadata({Metadata::Real}));
   num_active_ = 0;
-  max_active_index_ = 0;
+  max_active_index_ = -1;
```
Comment (brryan) [inline src/interface/swarm.cpp]: This was a bug

File: src/mesh/forest/tree.cpp [hunk 2]
```diff
@@ -321,11 +321,13 @@ Tree::GetBlockBCs(const LogicalLocation &loc) const {
 }
 
 void Tree::AddNeighborTree(CellCentOffsets offset, std::shared_ptr<Tree> neighbor_tree,
-                           RelativeOrientation orient) {
+                           RelativeOrientation orient, const bool periodic) {
```
Comment (brryan) [inline src/mesh/forest/tree.cpp:324]: @lroberts36 Here is the change to Forests/Trees we discussed (really just the solution you provided) so particles can discriminate between block-block connectivity internal to the domain or across periodic boundaries.

File: src/bvals/boundary_conditions_generic.hpp [hunk 3]
```diff
@@ -35,7 +35,103 @@ namespace parthenon {
...
+  auto &y = swarm->Get<Real>("y").Get();
+  auto &z = swarm->Get<Real>("z").Get();
+
+  // TODO(BRR) do something about all these if statements
+  pmb->par_for(
+      PARTHENON_AUTO_LABEL, 0, max_active_index, KOKKOS_LAMBDA(const int n) {
+        if (swarm_d.IsActive(n)) {
+          if (X1) {
+            if (INNER) {
+              if (TYPE == BCType::Periodic) {
```
Comment (Yurlungur) [inline src/bvals/boundary_conditions_generic.hpp]: I think because it's a template parameter you can make these all `if constexpr` and then the if statements are just fine.

File: src/mesh/mesh.cpp [hunk 4]
```diff
@@ -1185,10 +1205,32 @@ std::shared_ptr<MeshBlock> Mesh::FindMeshBlock(int tgid) const {
...
+      block_bcs[GetInnerBoundaryFace(dir)] =
+          loc.l(dir - 1) == 0 ? mesh_bcs[GetInnerBoundaryFace(dir)] : BoundaryFlag::block;
+      block_bcs[GetOuterBoundaryFace(dir)] = loc.l(dir - 1) == nrbx_ll - 1
+                                                 ? mesh_bcs[GetOuterBoundaryFace(dir)]
+                                                 : BoundaryFlag::block;
+    } else {
+      block_bcs[GetInnerBoundaryFace(dir)] = mesh_bcs[GetInnerBoundaryFace(dir)];
+      block_bcs[GetOuterBoundaryFace(dir)] = mesh_bcs[GetOuterBoundaryFace(dir)];
+    }
+  }
```
Comment (Yurlungur) [inline src/mesh/mesh.cpp]: After you get this in maybe it's time for me to return to #994 :D

File: example/particle_tracers/particle_tracers.cpp [hunk 5]
```diff
@@ -515,7 +513,7 @@ TaskCollection ParticleDriver::MakeTaskCollection(BlockList_t &blocks, int stage
       auto deposit = tl.AddTask(receive, tracers_example::DepositTracers, pmb.get());
 
       // Defragment if swarm memory pool occupancy is 90%
-      auto defrag = tl.AddTask(none, &SwarmContainer::Defrag, sc.get(), 0.9);
+      auto defrag = tl.AddTask(deposit, &SwarmContainer::Defrag, sc.get(), 0.9);
```
Comment (Yurlungur) [inline example/particle_tracers/particle_tracers.cpp:511]: is this due to a change in swarms or is this a bug?

**Review by Yurlungur — COMMENTED** [review]
:+1 :+1: :+1: I'm super happy to see this. This is a huge cleanup for readability, performance, and portability since it enables particles on HIP.

I'm a little confused though, I don't see where physical boundary conditions are now applied.

File: example/particle_tracers/particle_tracers.cpp (hunk 5 above)
Comment (brryan) [reply example/particle_tracers/particle_tracers.cpp:511]: bug :(

File: src/mesh/mesh.cpp (hunk 4 above)
Comment (brryan) [reply src/mesh/mesh.cpp]: Yeah I wonder if there's some more logic to think through about automatically figuring out how to disentangle different BCs for fields and particles as well.

File: src/bvals/boundary_conditions_generic.hpp (hunk 3 above)
Comment (brryan) [reply src/bvals/boundary_conditions_generic.hpp]: Made that change (in the fields version of the generic BC as well) thanks for the suggestion I think `if constexpr` should work nicely here, the alternative would be verbose.

File: src/interface/swarm_container.cpp [hunk 6]
```diff
@@ -160,6 +187,8 @@ TaskStatus SwarmContainer::Receive(BoundaryCommSubset phase) {
   for (auto &s : swarmVector_) {
     if (s->Receive(phase)) {
       success++;
+      ApplySwarmBoundaryConditions(s);
```
Comment (brryan) [inline src/interface/swarm_container.cpp:200]: physical boundaries are applied here, as part of `SwarmContainer::Receive`. Maybe this could be made a little more obvious by renaming this function `ReceiveAndApplyBoundaries` or something. In general boundaries *must* be applied after send/receives to handle periodic boundary conditions so that was my motivation for calling `ApplySwarmBoundaryConditions` here.

File: src/mesh/forest/tree.cpp (hunk 2 above)
Comment (lroberts36) [reply src/mesh/forest/tree.cpp:324]: That looks good for now. I am making a few changes to the tree boundary conditions in #1047 that should remove the need for this. When that is ready to go, it would probably be good to get your eyes on it since it moves around a bit where boundary conditions and BC function pointers are stored.

File: src/bvals/boundary_conditions_generic.hpp (hunk 3 above)
Comment (brryan) [reply src/bvals/boundary_conditions_generic.hpp]: I was not able to switch the field version of `GenericBC` to use `constexpr if` -- the nvcc compiler errors ultimately weren't making sense so I'm worried the compiler just isn't stable enough to support `constexpr if` on a function that intricate. Maybe that function can be revisited in a subsequent PR.

File: src/interface/swarm.cpp (hunk 1 above)
Comment (brryan) [reply src/interface/swarm.cpp]: This is a magic number so maybe I need a `constexpr int Swarm::inactive_max_index = -1;`

File: src/bvals/boundary_conditions.hpp [hunk 7]
```diff
@@ -43,6 +42,12 @@ TaskStatus ApplyBoundaryConditionsMD(std::shared_ptr<MeshData<Real>> &pmd);
 TaskStatus ApplyBoundaryConditionsOnCoarseOrFineMD(std::shared_ptr<MeshData<Real>> &pmd,
                                                    bool coarse);
 
+inline TaskStatus ApplyBoundaryConditions(std::shared_ptr<MeshBlockData<Real>> &rc);
```
Comment (lroberts36) [inline src/bvals/boundary_conditions.hpp]: Why is this declared again here when it is defined a few lines above? I must be missing something.

File: src/interface/swarm.cpp (hunk 1 above)
Comment (lroberts36) [reply src/interface/swarm.cpp]: yeah, defining a constant seems clearer.

File: src/interface/swarm_device_context.hpp [hunk 8]
```diff
@@ -60,10 +60,15 @@ class SwarmDeviceContext {
...
+    // Particle is on neither this block nor a neighboring block
     if (i < 0 || i > 3 || ((j < 0 || j > 3) && ndim_ > 1) ||
         ((k < 0 || k > 3) && ndim_ > 2)) {
-      PARTHENON_FAIL("Particle neighbor indices out of bounds");
+      printf("[%i] k = %i j = %i i = %i\n", n, k, j, i);
+      printf("x = %e [%e %e]\n", x, x_min_, x_max_);
+      printf("y = %e [%e %e]\n", y, y_min_, y_max_);
+      printf("z = %e [%e %e]\n", z, z_min_, z_max_);
+      PARTHENON_FAIL("Particle neighbor indices out of bounds; particle has somehow "
+                     "moved beyond the halo of adjacent blocks which is not permitted.");
```
Comment (lroberts36) [inline src/interface/swarm_device_context.hpp:71]: Are these print statements for testing or should they stay?

File: src/mesh/mesh.cpp [hunk 9]
```diff
@@ -1213,10 +1233,32 @@ std::shared_ptr<MeshBlock> Mesh::FindMeshBlock(int tgid) const {
...
```
Comment (lroberts36) [inline src/mesh/mesh.cpp]: Are these changes due to a merge issue? They look like old code that got removed in one of the forest PRs.

**Review by lroberts36 — COMMENTED** [review]
This is a nice refactor that LGTM, except for the one issue possible issue I mention below with the `SetBlockSizeAndBoundaries`. Once that is resolved I am happy to approve. 

I think there will be some work to get this into #1047, since things like `Mesh::mesh_bcs` are now attached to trees instead of the entire mesh. I will wait on making that non-WIP until this is in develop and I can merge it into that branch.

File: src/interface/swarm_device_context.hpp (hunk 8 above)
Comment (brryan) [reply src/interface/swarm_device_context.hpp:71]: This is a unique routine in that when anything goes wrong in particle transport in downstream applications, it is almost always caught by this check. So I wanted to make the output unusually verbose to speed up debugging.

File: src/bvals/boundary_conditions.hpp (hunk 7 above)
Comment (brryan) [reply src/bvals/boundary_conditions.hpp]: Not sure how that happened, thanks for noticing this, removed.

File: src/interface/swarm.cpp (hunk 1 above)
Comment (brryan) [reply src/interface/swarm.cpp]: Switched to `Swarm::inactive_max_active_index_`

File: src/interface/swarm_device_context.hpp (hunk 8 above)
Comment (brryan) [reply src/interface/swarm_device_context.hpp:71]: Also, I'm using `printf` here because this is device code.

File: src/mesh/mesh.cpp (hunk 9 above)
Comment (brryan) [reply src/mesh/mesh.cpp]: Yeah I agree I don't know how this old code got back into this merge, I mistakenly thought it was part of the fix for periodic vs block BCs but it isn't I don't think. `Mesh::SetBlockSizeAndBoundaries` should be unchanged now, thanks for noticing this.

Comment (lroberts36) [reply src/mesh/mesh.cpp]: yeah, weird. Sometimes git does strange things...

**Review by lroberts36 — APPROVED** [review]
LGTM

**Review by Yurlungur — APPROVED** [review]
LGTM. Thanks for the cleanup @brryan !

File: src/mesh/mesh.cpp (hunk 4 above)
Comment (Yurlungur) [reply src/mesh/mesh.cpp]: I think my previous PR should help with that. I'll revive it when I have the time.

File: src/bvals/boundary_conditions_generic.hpp (hunk 3 above)
Comment (Yurlungur) [reply src/bvals/boundary_conditions_generic.hpp]: seems like this is resolved.

**Comment by lroberts36** [comment]
@brryan @Yurlungur Is there any reason not to put this on automerge now?

**Comment by Yurlungur** [comment]
> @brryan @Yurlungur Is there any reason not to put this on automerge now?

No objections from me.

**Comment by brryan** [comment]
> @brryan @Yurlungur Is there any reason not to put this on automerge now?

I was going to just try this downstream later this week to check for oversights. Let me also add Philipp as a reviewer and give him a chance to look at this if he wants to.

**Comment by lroberts36** [comment]
@brryan: ok, that makes sense. I was just going to try and start seeing what is required to merge this with the mesh constructor refactor PR and the forest PR that is currently in WIP.

**Comment by pgrete** [comment]
As for the SwarmPacks I'm happy with the discussion and don't plan a detailed review. So just go ahead an merge when everyone is happy.

**Comment by brryan** [comment]
I verified that this works downstream, including custom reflecting BCs for particles, so I'm enabling auto-merge now. Thanks everyone for reviews
